compiler/rustc_trait_selection/src/traits/engine.rs

```diff
@@ -1,18 +1,173 @@
-use rustc_middle::ty::TyCtxt;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_middle::ty::{FoundRelationships, Ty, TyCtxt, TyVid};
 
 use super::TraitEngine;
 use super::{ChalkFulfillmentContext, FulfillmentContext};
 
+use std::panic;
+
 pub trait TraitEngineExt<'tcx> {
     fn new(tcx: TyCtxt<'tcx>) -> Box<Self>;
 }
 
 impl<'tcx> TraitEngineExt<'tcx> for dyn TraitEngine<'tcx> {
     fn new(tcx: TyCtxt<'tcx>) -> Box<Self> {
-        if tcx.sess.opts.debugging_opts.chalk {
+        if tcx.sess.opts.debugging_opts.chalk_migration {
+            Box::new(ChalkMigration {
+                chalk: ChalkFulfillmentContext::new(),
+                legacy: FulfillmentContext::new(),
+            })
+        } else if tcx.sess.opts.debugging_opts.chalk {
             Box::new(ChalkFulfillmentContext::new())
         } else {
             Box::new(FulfillmentContext::new())
         }
     }
 }
+
+struct ChalkMigration<'tcx> {
+    chalk: ChalkFulfillmentContext<'tcx>,
+    legacy: FulfillmentContext<'tcx>,
+}
+
+macro_rules! try_chalk {
+    ($chalk:expr, $legacy:expr, $method:ident, $($args:expr),*) => {{
+        let legacy = $legacy.$method($($args.clone()),*);
+        let hook = panic::take_hook();
+        panic::set_hook(Box::new(|_| {
+            // report nothing
+        }));
+        match panic::catch_unwind(panic::AssertUnwindSafe(|| {
+            let chalk = $chalk.$method($($args),*);
+            if chalk != legacy {
+                chalk.report(&legacy);
+            }
+        })) {
+            Ok(()) => {},
+            Err(_) => eprintln!("chalk panicked, rerun with `-Zchalk` if you are interested in the ICE itself."),
+        }
+        panic::set_hook(hook);
+        legacy
+    }};
+}
+
+trait ChalkDiff {
+    fn report(&self, legacy: &Self);
+}
+
+impl<'tcx> ChalkDiff for Ty<'tcx> {
```

Comment:
Ugh these impls are...kind of hacky. Maybe instead the `try_chalk` macro could be passed a closure?

Reply:
😄 yea, that seems better